Multi-select fields are breaking some APIs like `getOrganizationTickets(orgId)`

The introduction of Multi-select fields  in March 2017 in beta and recently in production is breaking some APIs like `getOrganizationTickets(orgId)`

https://support.zendesk.com/hc/en-us/articles/360000032087-New-ticket-field-type-Multi-select

```
org.zendesk.client.v2.ZendeskException: java.lang.IllegalArgumentException: Can not deserialize instance of java.lang.String out of START_ARRAY token
 at [Source: N/A; line: -1, column: -1] (through reference chain: org.zendesk.client.v2.model.Ticket["custom_fields"]->org.zendesk.client.v2.model.CustomFieldValue["value"])

	at org.zendesk.client.v2.Zendesk.complete(Zendesk.java:2139)
	at org.zendesk.client.v2.Zendesk.access$1400(Zendesk.java:91)
	at org.zendesk.client.v2.Zendesk$PagedIterable$PagedIterator.hasNext(Zendesk.java:2341)
	at org.zendesk.client.v2.RealSmokeTest.getOrganizationTickets(RealSmokeTest.java:238)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:160)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: Can not deserialize instance of java.lang.String out of START_ARRAY token
 at [Source: N/A; line: -1, column: -1] (through reference chain: org.zendesk.client.v2.model.Ticket["custom_fields"]->org.zendesk.client.v2.model.CustomFieldValue["value"])
	at com.ning.http.client.providers.netty.future.NettyResponseFuture.done(NettyResponseFuture.java:220)
	at com.ning.http.client.providers.netty.handler.HttpProtocol.finishUpdate(HttpProtocol.java:194)
	at com.ning.http.client.providers.netty.handler.HttpProtocol.handleChunk(HttpProtocol.java:451)
	at com.ning.http.client.providers.netty.handler.HttpProtocol.handle(HttpProtocol.java:474)
	at com.ning.http.client.providers.netty.handler.Processor.messageReceived(Processor.java:88)
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
	at org.jboss.netty.handler.stream.ChunkedWriteHandler.handleUpstream(ChunkedWriteHandler.java:142)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
	at org.jboss.netty.handler.codec.http.HttpContentDecoder.messageReceived(HttpContentDecoder.java:135)
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296)
	at org.jboss.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:452)
	at org.jboss.netty.handler.codec.replay.ReplayingDecoder.callDecode(ReplayingDecoder.java:536)
	at org.jboss.netty.handler.codec.replay.ReplayingDecoder.messageReceived(ReplayingDecoder.java:435)
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at org.jboss.netty.handler.codec.http.HttpClientCodec.handleUpstream(HttpClientCodec.java:92)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296)
	at org.jboss.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:462)
	at org.jboss.netty.handler.codec.frame.FrameDecoder.callDecode(FrameDecoder.java:443)
	at org.jboss.netty.handler.codec.frame.FrameDecoder.messageReceived(FrameDecoder.java:310)
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:559)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:268)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:255)
	at org.jboss.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88)
	at org.jboss.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:108)
	at org.jboss.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:337)
	at org.jboss.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:89)
	at org.jboss.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178)
	at org.jboss.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108)
	at org.jboss.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalArgumentException: Can not deserialize instance of java.lang.String out of START_ARRAY token
 at [Source: N/A; line: -1, column: -1] (through reference chain: org.zendesk.client.v2.model.Ticket["custom_fields"]->org.zendesk.client.v2.model.CustomFieldValue["value"])
	at com.fasterxml.jackson.databind.ObjectMapper._convert(ObjectMapper.java:2518)
	at com.fasterxml.jackson.databind.ObjectMapper.convertValue(ObjectMapper.java:2445)
	at org.zendesk.client.v2.Zendesk$PagedAsyncListCompletionHandler.onCompleted(Zendesk.java:1945)
	at org.zendesk.client.v2.Zendesk$PagedAsyncListCompletionHandler.onCompleted(Zendesk.java:1929)
	at com.ning.http.client.AsyncCompletionHandler.onCompleted(AsyncCompletionHandler.java:55)
	at com.ning.http.client.providers.netty.future.NettyResponseFuture.getContent(NettyResponseFuture.java:177)
	at com.ning.http.client.providers.netty.future.NettyResponseFuture.done(NettyResponseFuture.java:214)
	... 41 more
Caused by: com.fasterxml.jackson.databind.JsonMappingException: Can not deserialize instance of java.lang.String out of START_ARRAY token
 at [Source: N/A; line: -1, column: -1] (through reference chain: org.zendesk.client.v2.model.Ticket["custom_fields"]->org.zendesk.client.v2.model.CustomFieldValue["value"])
	at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:164)
	at com.fasterxml.jackson.databind.DeserializationContext.mappingException(DeserializationContext.java:599)
	at com.fasterxml.jackson.databind.deser.std.StringDeserializer.deserialize(StringDeserializer.java:41)
	at com.fasterxml.jackson.databind.deser.std.StringDeserializer.deserialize(StringDeserializer.java:11)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:375)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:98)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:308)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:121)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:226)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:203)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:23)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:375)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:98)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:308)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:121)
	at com.fasterxml.jackson.databind.ObjectMapper._convert(ObjectMapper.java:2513)
	... 47 more

```

cc @alecharp 
FYI @stephenc (I imagine we still don't have a real test environment provided by @zendesk ?)

--- a/pom.xml
+++ b/pom.xml
@@ -85,6 +85,18 @@
     <url>https://opensource.ci.cloudbees.com/job/zendesk-java-client/</url>
   </ciManagement>
 
+  <dependencyManagement>
+    <dependencies>
+      <dependency>
+        <groupId>com.fasterxml.jackson</groupId>
+        <artifactId>jackson-bom</artifactId>
+        <version>2.9.3</version>
+        <type>pom</type>
+        <scope>import</scope>
+      </dependency>
+    </dependencies>
+  </dependencyManagement>
+
   <dependencies>
     <dependency>
       <groupId>org.slf4j</groupId>
@@ -99,12 +111,10 @@
     <dependency>
       <groupId>com.fasterxml.jackson.core</groupId>
       <artifactId>jackson-core</artifactId>
-      <version>2.9.3</version>
     </dependency>
     <dependency>
       <groupId>com.fasterxml.jackson.core</groupId>
       <artifactId>jackson-databind</artifactId>
-      <version>2.9.3</version>
     </dependency>
     <dependency>
       <groupId>com.damnhandy</groupId>

--- a/pom.xml
+++ b/pom.xml
@@ -99,12 +99,12 @@
     <dependency>
       <groupId>com.fasterxml.jackson.core</groupId>
       <artifactId>jackson-core</artifactId>
-      <version>2.1.4</version>
+      <version>2.9.3</version>
     </dependency>
     <dependency>
       <groupId>com.fasterxml.jackson.core</groupId>
       <artifactId>jackson-databind</artifactId>
-      <version>2.1.4</version>
+      <version>2.9.3</version>
     </dependency>
     <dependency>
       <groupId>com.damnhandy</groupId>

--- a/src/main/java/org/zendesk/client/v2/model/Action.java
+++ b/src/main/java/org/zendesk/client/v2/model/Action.java
@@ -1,35 +1,26 @@
 package org.zendesk.client.v2.model;
 
-import com.fasterxml.jackson.core.JsonParser;
-import com.fasterxml.jackson.databind.DeserializationContext;
-import com.fasterxml.jackson.databind.JsonDeserializer;
-import com.fasterxml.jackson.databind.JsonNode;
-import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.annotation.JsonFormat;
 
-import java.io.IOException;
-import java.util.ArrayList;
 import java.util.Arrays;
-import java.util.Collections;
-import java.util.List;
 
 /**
- * Zendesk sometimes returns JSON with a String instead of String[] so we need a custom deserializer.
- *
  * @author adavidson
  * @author Johno Crawford (johno@sulake.com)
  */
-@JsonDeserialize(using = Action.ActionDeserializer.class)
 public class Action {
 
     private String field;
-    private List<String> value;
+
+    @JsonFormat(with = JsonFormat.Feature.ACCEPT_SINGLE_VALUE_AS_ARRAY)
+    private String[] value;
 
     public Action() {
     }
 
-    public Action(String field, String... values) {
+    public Action(String field, String[] value) {
         this.field = field;
-        this.value = Arrays.asList(values);
+        this.value = value;
     }
 
     public String getField() {
@@ -40,44 +31,19 @@ public class Action {
         this.field = field;
     }
 
-    public List<String> getValue() {
+    public String[] getValue() {
         return value;
     }
 
-    public void setValue(List<String> value) {
+    public void setValue(String[] value) {
         this.value = value;
     }
 
     @Override
     public String toString() {
-        final StringBuilder sb = new StringBuilder();
-        sb.append("Action");
-        sb.append("{field=").append(field);
-        sb.append(", value=").append(value);
-        sb.append('}');
-        return sb.toString();
+        return "Action{" +
+                "field='" + field + '\'' +
+                ", value=" + Arrays.toString(value) +
+                '}';
     }
-
-    static class ActionDeserializer extends JsonDeserializer<Action> {
-
-        @Override
-        public Action deserialize(JsonParser jsonParser, DeserializationContext context) throws IOException {
-            JsonNode root = jsonParser.getCodec().readTree(jsonParser);
-            Action action = new Action();
-            action.field = root.get("field").asText();
-
-            JsonNode valueNode = root.get("value");
-            if (valueNode.isArray()) {
-                List<String> values = new ArrayList<String>(valueNode.size());
-                for (final JsonNode node : valueNode) {
-                    values.add(node.asText());
-                }
-                action.value = values;
-            } else {
-                action.value = Collections.singletonList(valueNode.asText());
-            }
-            return action;
-        }
-    }
-
 }

--- a/src/main/java/org/zendesk/client/v2/model/CustomFieldValue.java
+++ b/src/main/java/org/zendesk/client/v2/model/CustomFieldValue.java
@@ -46,8 +46,9 @@ public class CustomFieldValue implements Serializable {
 
     @Override
     public String toString() {
-        return "CustomFieldValue{" + "id=" + id +
-                ", value=" + (value == null ? "null" : Arrays.asList(value).toString()) +
+        return "CustomFieldValue{" +
+                "id=" + id +
+                ", value=" + Arrays.toString(value) +
                 '}';
     }
 }

--- a/src/main/java/org/zendesk/client/v2/model/CustomFieldValue.java
+++ b/src/main/java/org/zendesk/client/v2/model/CustomFieldValue.java
@@ -1,7 +1,9 @@
 package org.zendesk.client.v2.model;
 
 import java.io.Serializable;
+import java.util.Arrays;
 
+import com.fasterxml.jackson.annotation.JsonFormat;
 import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
 
 /**
@@ -14,12 +16,14 @@ public class CustomFieldValue implements Serializable {
     private static final long serialVersionUID = 1L;
 
     private Long id;
-    private String value;
+
+    @JsonFormat(with = JsonFormat.Feature.ACCEPT_SINGLE_VALUE_AS_ARRAY)
+    private String[] value;
 
     public CustomFieldValue() {
     }
 
-    public CustomFieldValue(Long id, String value) {
+    public CustomFieldValue(Long id, String[] value) {
         this.id = id;
         this.value = value;
     }
@@ -32,11 +36,18 @@ public class CustomFieldValue implements Serializable {
         this.id = id;
     }
 
-    public String getValue() {
+    public String[] getValue() {
         return value;
     }
 
-    public void setValue(String value) {
+    public void setValue(String[] value) {
         this.value = value;
+    }
+
+    @Override
+    public String toString() {
+        return "CustomFieldValue{" + "id=" + id +
+                ", value=" + (value == null ? "null" : Arrays.asList(value).toString()) +
+                '}';
     }
 }

--- a/src/test/java/org/zendesk/client/v2/model/ActionTest.java
+++ b/src/test/java/org/zendesk/client/v2/model/ActionTest.java
@@ -1,0 +1,39 @@
+package org.zendesk.client.v2.model;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+
+public class ActionTest {
+
+    private Action parseJson(byte[] json) {
+        ObjectMapper mapper = new ObjectMapper();
+        try {
+            return mapper.readValue(json, Action.class);
+        } catch (Exception e) {
+            e.printStackTrace();
+            return null;
+        }
+    }
+
+    @Test
+    public void testActionWithSingleValue() throws Exception {
+        String json = "{ \"field\": 21337631753, \"value\": \"huuhaa\" }";
+        Action action = parseJson(json.getBytes());
+        assertNotNull(action);
+        assertEquals(1, action.getValue().length);
+        assertEquals("huuhaa", action.getValue()[0]);
+    }
+
+    @Test
+    public void testActionWithValues() throws Exception {
+        String json = "{ \"field\": 21337631753, \"value\": [\"huu\", \"haa\"] }";
+        Action action = parseJson(json.getBytes());
+        assertNotNull(action);
+        assertEquals(2, action.getValue().length);
+        assertEquals("huu", action.getValue()[0]);
+        assertEquals("haa", action.getValue()[1]);
+    }
+}

--- a/src/test/java/org/zendesk/client/v2/model/EventTest.java
+++ b/src/test/java/org/zendesk/client/v2/model/EventTest.java
@@ -21,7 +21,7 @@ public class EventTest {
     try {
       return mapper.readValue( json, Event.class );
     } catch ( Exception e ) {
-      System.out.println( e );
+      e.printStackTrace();
       return null;
     }
   }

--- a/src/test/java/org/zendesk/client/v2/model/TicketResultTest.java
+++ b/src/test/java/org/zendesk/client/v2/model/TicketResultTest.java
@@ -13,7 +13,7 @@ public class TicketResultTest {
         try {
             return mapper.readValue(json, TicketResult.class);
         } catch (Exception e) {
-            System.out.println(e);
+            e.printStackTrace();
             return null;
         }
     }

